ComInterfaceGenerator generates code with pointer type in generic

In custom collections generated code, the code generated for the following snippet contains `byte*` as a generic argument, causing a compiler error.

Input Code:
```C#
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.Marshalling;
[assembly:DisableRuntimeMarshalling]

[global::System.Runtime.InteropServices.Marshalling.GeneratedComInterfaceAttribute]
partial interface INativeAPI : IUnmanagedInterfaceType
{
    
    [return:MarshalUsing(ConstantElementCount=10)]
    System.Byte[] Method(
        System.Byte[] p,
        in System.Byte[] pIn,
        int pRefSize,
        [MarshalUsing(CountElementName = "pRefSize")] ref System.Byte[] pRef,
        [MarshalUsing(CountElementName = "pOutSize")] out System.Byte[] pOut,
        out int pOutSize);
}
// Try using the generated native interface
sealed class NativeAPI : IUnmanagedVirtualMethodTableProvider, INativeAPI.Native
{
    public VirtualMethodTableInfo GetVirtualMethodTableInfoForKey(System.Type type) => throw null;
}
partial interface INativeAPI
{
        static unsafe void* IUnmanagedInterfaceType.VirtualMethodTableManagedImplementation => null;
}
```
Generated Code:
```C#
// <auto-generated/>
unsafe partial interface INativeAPI
{
    internal unsafe partial interface Native
    {
        [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
        internal static ABI_Method(void* __this_native, byte* __p_native, byte** __pIn_native__param, int pRefSize, byte** __pRef_native__param, byte** __pOut_native__param, int* __pOutSize_native__param)
        {
            ...
            try
            {
                ...
            }
            catch (System.Exception __exception)
            {
                // error CS0306: The type 'byte*' may not be used as a type argument
                __retVal_native = System.Runtime.InteropServices.Marshalling.ExceptionDefaultMarshaller<byte*>.ConvertToUnmanaged(__exception);
            }
            finally
            {
                // Cleanup - Perform required cleanup.
               ...
            }

            return __retVal_native;
        }
    }
}
```

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ElementsMarshalling.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ElementsMarshalling.cs
@@ -396,7 +396,7 @@ namespace Microsoft.Interop
         /// This handles the case where the native type of a non-blittable managed type is a pointer,
         /// which are unsupported in generic type parameters.
         /// </summary>
-        private sealed class PointerNativeTypeAssignmentRewriter : CSharpSyntaxRewriter
+        internal sealed class PointerNativeTypeAssignmentRewriter : CSharpSyntaxRewriter
         {
             private readonly string _nativeIdentifier;
             private readonly PointerTypeSyntax _nativeType;

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ElementsMarshalling.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/ElementsMarshalling.cs
@@ -390,47 +390,5 @@ namespace Microsoft.Interop
 
             return EmptyStatement();
         }
-
-        /// <summary>
-        /// Rewrite assignment expressions to the native identifier to cast to IntPtr.
-        /// This handles the case where the native type of a non-blittable managed type is a pointer,
-        /// which are unsupported in generic type parameters.
-        /// </summary>
-        internal sealed class PointerNativeTypeAssignmentRewriter : CSharpSyntaxRewriter
-        {
-            private readonly string _nativeIdentifier;
-            private readonly PointerTypeSyntax _nativeType;
-
-            public PointerNativeTypeAssignmentRewriter(string nativeIdentifier, PointerTypeSyntax nativeType)
-            {
-                _nativeIdentifier = nativeIdentifier;
-                _nativeType = nativeType;
-            }
-
-            public override SyntaxNode VisitAssignmentExpression(AssignmentExpressionSyntax node)
-            {
-                if (node.Left.ToString() == _nativeIdentifier)
-                {
-                    return node.WithRight(
-                        CastExpression(MarshallerHelpers.SystemIntPtrType, node.Right));
-                }
-                if (node.Right.ToString() == _nativeIdentifier)
-                {
-                    return node.WithRight(CastExpression(_nativeType, node.Right));
-                }
-
-                return base.VisitAssignmentExpression(node);
-            }
-
-            public override SyntaxNode? VisitArgument(ArgumentSyntax node)
-            {
-                if (node.Expression.ToString() == _nativeIdentifier)
-                {
-                    return node.WithExpression(
-                        CastExpression(_nativeType, node.Expression));
-                }
-                return base.VisitArgument(node);
-            }
-        }
     }
 }

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/PointerNativeTypeAssignmentRewriter.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/PointerNativeTypeAssignmentRewriter.cs
@@ -1,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.Interop
+{
+    /// <summary>
+    /// Rewrite assignment expressions to the native identifier to cast to IntPtr.
+    /// This handles the case where the native type of a non-blittable managed type is a pointer,
+    /// which are unsupported in generic type parameters.
+    /// </summary>
+    internal sealed class PointerNativeTypeAssignmentRewriter : CSharpSyntaxRewriter
+    {
+        private readonly string _nativeIdentifier;
+        private readonly PointerTypeSyntax _nativeType;
+
+        public PointerNativeTypeAssignmentRewriter(string nativeIdentifier, PointerTypeSyntax nativeType)
+        {
+            _nativeIdentifier = nativeIdentifier;
+            _nativeType = nativeType;
+        }
+
+        public override SyntaxNode VisitAssignmentExpression(AssignmentExpressionSyntax node)
+        {
+            if (node.Left.ToString() == _nativeIdentifier)
+            {
+                return node.WithRight(
+                    CastExpression(MarshallerHelpers.SystemIntPtrType, node.Right));
+            }
+            if (node.Right.ToString() == _nativeIdentifier)
+            {
+                return node.WithRight(CastExpression(_nativeType, node.Right));
+            }
+
+            return base.VisitAssignmentExpression(node);
+        }
+
+        public override SyntaxNode? VisitArgument(ArgumentSyntax node)
+        {
+            if (node.Expression.ToString() == _nativeIdentifier)
+            {
+                return node.WithExpression(
+                    CastExpression(_nativeType, node.Expression));
+            }
+            return base.VisitArgument(node);
+        }
+    }
+}

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs
@@ -82,11 +82,19 @@ namespace Microsoft.Interop
             }
 
             // <nativeIdentifier> = <convertToUnmanaged>;
-            yield return ExpressionStatement(
-                AssignmentExpression(
+            var assignment = AssignmentExpression(
                     SyntaxKind.SimpleAssignmentExpression,
                     IdentifierName(nativeIdentifier),
-                    convertToUnmanaged));
+                    convertToUnmanaged);
+
+
+            if (_unmanagedType is PointerTypeInfo pointer)
+            {
+                var rewriter = new NonBlittableElementsMarshalling.PointerNativeTypeAssignmentRewriter(assignment.Right.ToString(), (PointerTypeSyntax)pointer.Syntax);
+                assignment = (AssignmentExpressionSyntax)rewriter.Visit(assignment);
+
+            }
+            yield return ExpressionStatement(assignment);
         }
 
         public IEnumerable<StatementSyntax> GeneratePinnedMarshalStatements(TypePositionInfo info, StubCodeContext context)

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs
@@ -90,7 +90,7 @@ namespace Microsoft.Interop
 
             if (_unmanagedType is PointerTypeInfo pointer)
             {
-                var rewriter = new NonBlittableElementsMarshalling.PointerNativeTypeAssignmentRewriter(assignment.Right.ToString(), (PointerTypeSyntax)pointer.Syntax);
+                var rewriter = new PointerNativeTypeAssignmentRewriter(assignment.Right.ToString(), (PointerTypeSyntax)pointer.Syntax);
                 assignment = (AssignmentExpressionSyntax)rewriter.Visit(assignment);
 
             }

